File: src/seni/Bind.js
/*
 *  Seni
 *  Copyright (C) 2015  Inderjit Gill <email@indy.io>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import Shapes from './Shapes';
import Paths from './Paths';
import MatrixStackBindings from './MatrixStackBindings';
import MathUtil from './MathUtil';
import ColourBindings from './ColourBindings';
import Core from './Core';
import Bracket from './BracketBindings';
import PseudoRandom from './PseudoRandom';
import Focal from './Focal';
import Repeat from './Repeat';

function createBind(env, pb, restArgs) {

  // call the PublicBinding's create function passing in an explicit self
  // along with any additional arguments
  let value = pb.create.apply(null, [pb].concat(restArgs));

  // bind the value to the pb's name
  return env.set(pb.name, { binding : value, pb : pb });
}

// applies the publicBindings in namespace to env
function applyPublicBindings(env, namespace) {
  // grab any additional arguments that have been given to this function
  let restArgs = Array.prototype.slice.call(arguments, 2);
  let bindings = namespace.publicBindings;

  return bindings.reduce((e, pb) => createBind(e, pb, restArgs), env);
}

const Bind = {
  addBindings: function(env, renderer) {
    env = applyPublicBindings(env, Core);
    env = applyPublicBindings(env, MathUtil);
    env = applyPublicBindings(env, PseudoRandom);
    env = applyPublicBindings(env, MatrixStackBindings, renderer);
    env = applyPublicBindings(env, Shapes, renderer);
    env = applyPublicBindings(env, Paths);
    env = applyPublicBindings(env, ColourBindings);
    env = applyPublicBindings(env, Focal, renderer);
    env = applyPublicBindings(env, Repeat, renderer);
    return env;
  },

  addBracketBindings: function(env, rng) {
    env = applyPublicBindings(env, Core);
    env = applyPublicBindings(env, MathUtil);
    env = applyPublicBindings(env, PseudoRandom);
    env = applyPublicBindings(env, ColourBindings);
    env = applyPublicBindings(env, Bracket, rng);

    return env;
  }
};

export default Bind;

File: core/sen-core/src/bind.rs
// Copyright (C) 2019 Inderjit Gill

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

use crate::compiler::Program;
use crate::error::{Error, Result};
use crate::vm::{Var, Vm};
use crate::keywords::Keyword;

pub fn bind_vector_length(vm: &mut Vm, _program: &Program, num_args: usize) -> Result<Var> {

    let mut vector: Option<&Var> = None;

    let mut args_pointer = vm.sp - (num_args * 2);

    for _ in 0..num_args {
        let label = &vm.stack[args_pointer + 0];
        let value = &vm.stack[args_pointer + 1];
        args_pointer += 2;

        if let Var::Int(iname) = label {
            if *iname == Keyword::Vector as i32 {
                vector = Some(value);
                break;
            }
        }
    }

    if let Some(v) = vector {
        if let Var::Vector(vs) = v {
            let len = vs.len();
            return Ok(Var::Int(len as i32))
        }
    }

    return Err(Error::Bind("bind_vector_length requires vector argument".to_string()))
}

#[cfg(test)]
mod tests {
    use crate::vm::tests::*;

    #[test]
    fn test_bind_vector_length() {
        is_int("(define v []) (++ v 100) (vector/length vector: v)", 1);
        is_int("(define v [1]) (++ v 100) (vector/length vector: v)", 2);
        is_int("(define v [1 2]) (++ v 100) (vector/length vector: v)", 3);
        is_int("(define v [1 2 3]) (++ v 100) (vector/length vector: v)", 4);
        is_int("(define v [1 2 3 4]) (++ v 100) (vector/length vector: v)", 5);
        is_int("(define v []) (++ v 4) (++ v 3) (++ v 2) (++ v 1) (++ v 0) (vector/length vector: v)", 5);
        is_int("(define v [1 2]) (++ v 98) (++ v 99) (++ v 100) (vector/length vector: v)", 5);
    }
}
